src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,283 @@
+    secp256k1_sha256 sha;
+    size_t i;
+
+    secp256k1_musig_keyagglist_sha256(&sha);
+    for (i = 0; i < np; i++) {
+        unsigned char ser[32];
+        if (!secp256k1_xonly_pubkey_serialize(ctx, ser, pk[i])) {
+            return 0;
+        }
+        secp256k1_sha256_write(&sha, ser, 32);
+    }
+    secp256k1_sha256_finalize(&sha, pk_hash);
+    return 1;
+}
+
+/* Initializes SHA256 with fixed midstate. This midstate was computed by applying
+ * SHA256 to SHA256("KeyAgg coefficient")||SHA256("KeyAgg coefficient"). */
+static void secp256k1_musig_keyaggcoef_sha256(secp256k1_sha256 *sha) {
+    secp256k1_sha256_initialize(sha);
+
+    sha->s[0] = 0x6ef02c5aul;
+    sha->s[1] = 0x06a480deul;
+    sha->s[2] = 0x1f298665ul;
+    sha->s[3] = 0x1d1134f2ul;
+    sha->s[4] = 0x56a0b063ul;
+    sha->s[5] = 0x52da4147ul;
+    sha->s[6] = 0xf280d9d4ul;
+    sha->s[7] = 0x4484be15ul;
+    sha->bytes = 64;
+}
+
+/* Compute KeyAgg coefficient which is constant 1 for the second pubkey and
+ * tagged_hash(pk_hash, x) where pk_hash is the hash of public keys otherwise.
+ * second_pk_x can be 0 in case there is no second_pk. Assumes both field
+ * elements x and second_pk_x are normalized. */
+static void secp256k1_musig_keyaggcoef_internal(secp256k1_scalar *r, const unsigned char *pk_hash, const secp256k1_fe *x, const secp256k1_fe *second_pk_x) {
+    secp256k1_sha256 sha;
+    unsigned char buf[32];
+
+    if (secp256k1_fe_cmp_var(x, second_pk_x) == 0) {
+        secp256k1_scalar_set_int(r, 1);
+    } else {
+        secp256k1_musig_keyaggcoef_sha256(&sha);
+        secp256k1_sha256_write(&sha, pk_hash, 32);
+        secp256k1_fe_get_b32(buf, x);
+        secp256k1_sha256_write(&sha, buf, 32);
+        secp256k1_sha256_finalize(&sha, buf);
+        secp256k1_scalar_set_b32(r, buf, NULL);
+    }
+
+}
+
+/* Assumes both field elements x and second_pk_x are normalized. */
+static void secp256k1_musig_keyaggcoef(secp256k1_scalar *r, const secp256k1_keyagg_cache_internal *cache_i, secp256k1_fe *x) {
+    secp256k1_musig_keyaggcoef_internal(r, cache_i->pk_hash, x, &cache_i->second_pk_x);
+}
+
+typedef struct {
+    const secp256k1_context *ctx;
+    /* pk_hash is the hash of the public keys */
+    unsigned char pk_hash[32];
+    const secp256k1_xonly_pubkey * const* pks;
+    secp256k1_fe second_pk_x;
+} secp256k1_musig_pubkey_agg_ecmult_data;
+
+/* Callback for batch EC multiplication to compute keyaggcoef_0*P0 + keyaggcoef_1*P1 + ...  */
+static int secp256k1_musig_pubkey_agg_callback(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *data) {
+    secp256k1_musig_pubkey_agg_ecmult_data *ctx = (secp256k1_musig_pubkey_agg_ecmult_data *) data;
+    int ret;
+    ret = secp256k1_xonly_pubkey_load(ctx->ctx, pt, ctx->pks[idx]);
+    /* pubkey_load can't fail because the same pks have already been loaded (and
+     * we test this) */
+    VERIFY_CHECK(ret);
+    secp256k1_musig_keyaggcoef_internal(sc, ctx->pk_hash, &pt->x, &ctx->second_pk_x);
```

Comment:
Is this really true?

We may have `break`ed the for loop that finds the second key.

Reply:
All pubkeys are loaded in `secp256k1_musig_compute_pk_hash`. I extended the comment to mention this.

Reply:
Oh sure.